sources/engine/Stride.Engine/Engine/Processors/InstancingProcessor.cs

```diff
@@ -0,0 +1,206 @@
+using System.Collections.Generic;
+using Stride.Core.Annotations;
+using Stride.Core.Mathematics;
+using Stride.Core.Threading;
+using Stride.Graphics;
+using Stride.Rendering;
+using Buffer = Stride.Graphics.Buffer;
+
+namespace Stride.Engine.Processors
+{
+    public class InstancingProcessor : EntityProcessor<InstancingComponent, InstancingProcessor.InstancingData>, IEntityComponentRenderProcessor
+    {
+        private readonly Dictionary<RenderModel, RenderInstancing> modelInstancingMap = new Dictionary<RenderModel, RenderInstancing>();
+        private ModelRenderProcessor modelRenderProcessor;
+
+        public class InstancingData
+        {
+            public TransformComponent TransformComponent;
+            public ModelComponent ModelComponent;
+            public RenderModel RenderModel;
+            public RenderInstancing RenderInstancing = new RenderInstancing();
+        }
+
+        public InstancingProcessor() 
+            : base (typeof(TransformComponent), typeof(ModelComponent)) // Requires TransformComponent and ModelComponent
+        {
+            // After TransformProcessor but before ModelRenderProcessor
+            Order = -100;
+        }
+
+        public VisibilityGroup VisibilityGroup { get; set; }
+
+        public override void Draw(RenderContext context)
+        {
+            // Process the components
+            Dispatcher.ForEach(ComponentDatas, item =>
+            {
+                UpdateInstancing(item.Key, item.Value);
+                TransferData(context, item.Key, item.Value.RenderInstancing);
+            });
+        }
+
+        private static void TransferData(RenderContext context, InstancingComponent instancingComponent, RenderInstancing renderInstancing)
+        {
+            var instancing = instancingComponent.Type;
+            renderInstancing.InstanceCount = instancingComponent.Enabled ? instancing.InstanceCount : 0;
+            renderInstancing.ModelTransformUsage = (int)instancing.ModelTransformUsage;
+
+            if (renderInstancing.InstanceCount > 0)
+            {
+                if (instancing is InstancingUserArray instancingUserArray)
+                {
+                    renderInstancing.BuffersManagedByUser = false;
+                    renderInstancing.WorldMatrices = instancingUserArray.WorldMatrices;
+                    renderInstancing.WorldInverseMatrices = instancingUserArray.WorldInverseMatrices;
+
+                    if ((renderInstancing.InstanceWorldBuffer == null || renderInstancing.InstanceWorldBuffer.ElementCount < instancing.InstanceCount))
+                    {
+                        renderInstancing.InstanceWorldBuffer?.Dispose();
+                        renderInstancing.InstanceWorldInverseBuffer?.Dispose();
+
+                        renderInstancing.InstanceWorldBuffer = CreateMatrixBuffer(context.GraphicsDevice, instancing.InstanceCount);
```

Comment:
Could probably gain an extremely minor perf gain using `renderInstancing.InstanceCount` instead of `instancing.InstanceCount` since they should be the same value at this point.

Reply:
Thx!